### PR TITLE
Modify S3394(COBOL): Migrate to LayC

### PR DESCRIPTION
**Pull request by loris-s-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S3394/cobol/rule.adoc [hunk 1]
```diff
@@ -1,21 +1,112 @@
...
+        IF USER-INPUT IS NUMERIC THEN
+            MOVE 'Y' TO WS-NUMERIC
+        ELSE
+            DISPLAY 'INVALID INPUT. PLEASE ENTER A NUMERIC VALUE.'
+        END-IF
+    END-PERFORM
+    STOP RUN.
 ----
 
+=== How does that work?
```
Comment (loris-s-sonarsource) [inline rules/S3394/cobol/rule.adoc:92]:
```suggestion
=== How does this work?
```

File: rules/S3394/cobol/rule.adoc [hunk 2]
```diff
@@ -24,8 +115,8 @@ This rule ignores uses of ``++ACCEPT FROM++`` with date/time-related inputs.
 
 == Resources
 
-* https://owasp.org/Top10/A03_2021-Injection/[OWASP Top 10 2021 Category A3] - Injection
-* https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[OWASP Top 10 2017 Category A1] - Injection
+* OWASP - https://owasp.org/Top10/A03_2021-Injection/[OWASP Top 10 2021 Category A3] - Injection
+* OWASP - https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[OWASP Top 10 2017 Category A1] - Injection
```
Comment (loris-s-sonarsource) [inline rules/S3394/cobol/rule.adoc:119]:
```suggestion
* OWASP - https://owasp.org/Top10/A03_2021-Injection/[Top 10 2021 Category A3] - Injection
* OWASP - https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[Top 10 2017 Category A1] - Injection
```

File: rules/S3394/cobol/rule.adoc [hunk 3]
```diff
@@ -1,21 +1,112 @@
...
+COBOL can lead to a number of security problems.
+
+==== Data integrity problems
+
+Without any form of validation, the ACCEPT keyword can store arbitrary input
+data, even if it does not conform to the expected format or type. This can lead
+to data corruption, causing the program to behave unpredictably, crash, or
+corrupt customer data.
+
+==== Intrusion attacks
```
Comment (loris-s-sonarsource) [inline rules/S3394/cobol/rule.adoc:25]:
```suggestion
==== Injection attacks
```

File: rules/S3394/cobol/rule.adoc [hunk 4]
```diff
@@ -1,21 +1,112 @@
+This vulnerability allows corrupted data to be stored and causes errors in
+parts of the code that are supposed to use this encrypted data.
+
 == Why is this an issue?
 
-The ``++ACCEPT++`` keyword does no editing or error checking of the data it stores, therefore its use can be dangerous. For this reason, ``++ACCEPT++`` should be avoided. 
+The `ACCEPT` statement causes data entered at the console or provided by the
+operating system to be made available to the program in a specific data
+element, without any validation or cleanup.
```
Comment (loris-s-sonarsource) [inline rules/S3394/cobol/rule.adoc:8]:
```suggestion
element, without any validation or sanitization.
```

File: rules/S3394/cobol/rule.adoc [hunk 5]
```diff
@@ -1,21 +1,112 @@
...
+To mitigate the risks associated with the ACCEPT keyword in COBOL, you should
+implement data validation and error checking, or directly avoid using `ACCEPT`
+altogether.
+
+==== Data Validation Routines
+
+Before using the ACCEPT keyword to store data, validate the data using custom
+routines. These routines should check if the data meets the expected format and
+type. For example, if you're expecting a date, the routine should check if the
+input data is indeed a date.
```
Comment (loris-s-sonarsource) [inline rules/S3394/cobol/rule.adoc:103]:
```suggestion
Before using the data coming from ACCEPT, validate the data using custom
routines. These routines should check if the data meets the expected format and
type. For example, if you're expecting a number, the routine should check if the
input data is indeed a number.
```

File: rules/S3394/cobol/rule.adoc [hunk 6]
```diff
@@ -1,21 +1,112 @@
...
+==== Noncompliant code example
+
+[source,cobol,diff-id=1,diff-type=noncompliant]
+----
+IDENTIFICATION DIVISION.
+PROGRAM-ID. EXAMPLE.
+
+DATA DIVISION.
+WORKING-STORAGE SECTION.
+01 USER-INPUT PIC X(4).
```
Comment (sebastien-andrivet-sonarsource) [inline rules/S3394/cobol/rule.adoc:52]: The program is invalid, so I suggest to make it valid:

01 WS-NUMERIC PIC X VALUE 'N'.
```suggestion
01 WS-NUMERIC PIC X VALUE 'N'.
01 USER-INPUT PIC X(4).
```

File: rules/S3394/cobol/rule.adoc [hunk 7]
```diff
@@ -24,8 +115,8 @@ This rule ignores uses of ``++ACCEPT FROM++`` with date/time-related inputs.
 
 == Resources
 
-* https://owasp.org/Top10/A03_2021-Injection/[OWASP Top 10 2021 Category A3] - Injection
-* https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[OWASP Top 10 2017 Category A1] - Injection
+* OWASP - https://owasp.org/Top10/A03_2021-Injection/[Top 10 2021 Category A3] - Injection
+* OWASP - https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[Top 10 2017 Category A1] - Injection
```
Comment (sebastien-andrivet-sonarsource) [inline rules/S3394/cobol/rule.adoc:119]: The links are not properly formatted.

```suggestion
* OWASP - https://owasp.org/Top10/A03_2021-Injection/[Top 10 2021 Category A3 - Injection]
* OWASP - https://owasp.org/www-project-top-ten/2017/A1_2017-Injection[Top 10 2017 Category A1 - Injection]
```

**Review by sebastien-andrivet-sonarsource — CHANGES_REQUESTED** [review]
A suggestion to fix the COBOL program. Two links are not properly formatted.

**Review by sebastien-andrivet-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**